src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs

```diff
@@ -657,8 +673,134 @@ public override Task WriteAsync(Byte[] array, int offset, int count, Cancellatio
+            // buffer the source stream creates in its CopyToAsync implementation.  As such, we halve the buffer size,
+            // to split it across the two.
+            bufferSize = Math.Max(1, bufferSize / 2);
+
+            // Do the copy
+            return new CopyToAsyncStream(this, destination, ArrayPool<byte>.Shared.Rent(bufferSize), cancellationToken).CopyFromSourceToDestination();
+        }
+
+        private sealed class CopyToAsyncStream : Stream
+        {
+            private readonly DeflateStream _deflateStream;
+            private readonly Stream _destination;
+            private readonly CancellationToken _cancellationToken;
+            private readonly byte[] _arrayPoolBuffer;
+            private int _arrayPoolBufferHighWaterMark;
+
+            public CopyToAsyncStream(DeflateStream deflateStream, Stream destination, byte[] arrayPoolBuffer, CancellationToken cancellationToken)
+            {
+                Debug.Assert(deflateStream != null);
+                Debug.Assert(destination != null);
+                Debug.Assert(arrayPoolBuffer != null && arrayPoolBuffer.Length > 0);
+
+                _deflateStream = deflateStream;
+                _destination = destination;
+                _arrayPoolBuffer = arrayPoolBuffer;
+                _cancellationToken = cancellationToken;
+            }
+
+            public async Task CopyFromSourceToDestination()
+            {
+                _deflateStream.IncrementAsyncOperations();
+                try
+                {
+                    // Flush any existing data in the inflater to the destination stream.
+                    while (true)
+                    {
+                        int bytesRead = _deflateStream._inflater.Inflate(_arrayPoolBuffer, 0, _arrayPoolBuffer.Length);
+                        if (bytesRead > 0)
+                        {
+                            if (bytesRead > _arrayPoolBufferHighWaterMark) _arrayPoolBufferHighWaterMark = bytesRead;
+                            await _destination.WriteAsync(_arrayPoolBuffer, 0, bytesRead, _cancellationToken).ConfigureAwait(false);
+                        }
+                        else break;
+                    }
+
+                    // Now, use the source stream's CopyToAsync to push directly to our inflater via this helper stream
+                    await _deflateStream._stream.CopyToAsync(this, _arrayPoolBuffer.Length, _cancellationToken).ConfigureAwait(false);
+                }
+                finally
+                {
+                    _deflateStream.DecrementAsyncOperations();
+
+                    Array.Clear(_arrayPoolBuffer, 0, _arrayPoolBufferHighWaterMark); // clear only the most we used
+                    ArrayPool<byte>.Shared.Return(_arrayPoolBuffer, clearArray: false);
+                }
+            }
+
+            public override async Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
```

Comment:
This instance gets passed as the destination argument to the wrapped stream's CopyToAsync, so that wrapped stream's CopyToAsync pushes data to this instance via this WriteAsync.  That's why this is a Stream ;)